mycroft/skills/mycroft_skill/mycroft_skill.py

```diff
@@ -166,19 +173,65 @@ def __init__(self, name=None, bus=None, use_settings=True):
         # Skill Public API
         self.public_api = {}
 
+        # New style skills can be fully initialized at once
+        # most existing skills however do not pass the required kwargs
+        # the advantage of this is that it makes __init__ method safer
+        # for skill developers as long as they call super() before their code
+        if bus and skill_id:
+            self._startup(bus, skill_id)
+
     @property
-    def dialog_renderer(self):
-        if self.lang in self.dialog_renderers:
-            return self.dialog_renderers[self.lang]
-        # Try to load the renderer
-        self._load_dialog_files(self.root_dir, self.lang)
-        if self.lang in self.dialog_renderers:
-            return self.dialog_renderers[self.lang]
-        # Fall back to main language
-        return self.dialog_renderers.get(self._core_lang)
+    def is_fully_initialized(self):
+        return self._init_event.is_set()
+
+    def _startup(self, bus, skill_id=""):
+        """Startup the skill.
+
+        This connects the skill to the messagebus, loads vocabularies and
+        data files and in the end calls the skill creator's "intialize" code.
+
+        Arguments:
+            bus: Mycroft Messagebus connection object.
+            skill_id (str): need to be unique, by default is set from skill path
+                but skill loader can override this
+        """
+        if self.is_fully_initialized:
+            return  # already initialized
+        # NOTE: this method is called by SkillLoader
+        # it is private to make it clear to skill devs they should not touch it
+        try:
+            # set the skill_id
+            self.skill_id = skill_id or basename(self.root_dir)
+            self._init_settings()
+            self._init_filesystem()
+
+            # initialize anything that depends on the messagebus
+            self.bind(bus)
+            self.load_data_files()
+            self._register_decorated()
+            self.register_resting_screen()
+
+            # run skill developer initialization code
+            self.initialize()
+            self._check_for_first_run()
+            self._init_event.set()
+        except Exception as e:
+            LOG.exception('Skill initialization failed')
+            # If an exception occurs, attempt to clean up the skill
+            try:
+                self.default_shutdown()
+            except Exception as e2:
+                pass
+            raise e
 
     def _init_settings(self):
         """Setup skill settings."""
+        LOG.debug(f"initializing skill settings for {self.skill_id}")
+
+        # if settings were used in __init__ method self._startup won't have been called yet
+        # assume these are default values
+        if self.settings:
+            self._initial_settings = dict(self.settings)
```

Comment:
the reason i originally moved away from using deepcopy was performance

any technical advantage over using copy instead of dict? 

If it's about readability i personally find dict more readable since it uses only builtin syntax of the language, there is some "mental overhead" to figure out what copy function is being used that just isn't there when using dict

Reply:
I'm the opposite since dict always makes me pause, while copy() makes it clear what the intention is.

In performance from some benchmarking copy is ~4 times faster.

Reply:
will do the change :+1: